builtin/logical/aws/client.go

```diff
@@ -51,6 +53,8 @@ func getRootConfig(s logical.Storage) (*aws.Config, error) {
 	return &aws.Config{
 		Credentials: creds,
 		Region:      aws.String(credsConfig.Region),
+		IAMEndpoint: aws.String(credsConfig.IAMEndpoint),
+		STSEndpoint: aws.String(credsConfig.STSEndpoint),
```

Comment:
Pretty sure the issue is because you're using the upstream [`aws.Config`](https://docs.aws.amazon.com/sdk-for-go/api/aws/#Config) object and it only has an `Endpoint` member. Consider passing a `clientType` parameter to `getRootConfig` and if `clientType` is `iam` then you set `Endpoint` to `credsConfig.IAMEndpoint` and if it's `sts` then you set `Endpoint` to `credsConfig.STSEndpoint`. Largely the same code lives in the auth backend:
 https://github.com/hashicorp/vault/blob/1c4baa56e9882449ed70c0021100336a3465ea58/builtin/credential/aws/client.go#L24-L68